x/mongo/driver/topology/fsm.go

```diff
@@ -228,30 +228,75 @@ func (f *fsm) checkIfHasPrimary() {
 	}
 }
 
-func (f *fsm) updateRSFromPrimary(s description.Server) {
-	if f.SetName == "" {
-		f.SetName = s.SetName
-	} else if f.SetName != s.SetName {
-		f.removeServerByAddr(s.Addr)
-		f.checkIfHasPrimary()
-		return
+// hasStalePrimary returns true if the topology has a primary that is stale.
+func hasStalePrimary(fsm fsm, srv description.Server) bool {
+	if srv.WireVersion == nil {
+		return true
+	}
+
+	// Compare the election ID values of the server and the topology lexicographically.
+	compRes := bytes.Compare(srv.ElectionID[:], fsm.maxElectionID[:])
+
+	if srv.WireVersion.Max < 17 {
+		return compRes == -1 || fsm.maxSetVersion > srv.SetVersion
+	}
+
+	return compRes != 1 && (compRes != 0 || srv.SetVersion < fsm.maxSetVersion)
```

Comment:
```suggestion
        // Post 6.0, return true (stale primary) if electionID is less than maxElectionID or electionID
        // is equal to maxElectionID but setVersion is less than maxSetVersion.
	return compRes == -1 || (compRes == 0 && srv.SetVersion < fsm.maxSetVersion)
```
Similarly a comment and slight reordering of the conditional here may help with readability. I _think_ the conditional I've written above is identical to yours, but I could be wrong 🧑‍🔧 .

Reply:
What you have is not the same thing. The piecewise function that is used to compare looks like this:

<img width="635" alt="Screen Shot 2022-10-17 at 9 56 26 AM" src="https://user-images.githubusercontent.com/24281431/196225374-0159cb53-3f6b-4460-b014-dc307356fb4d.png">

So the original logic would be

```
a <= b && (a != b || c < d)
```

Your change would be

```
a < b || (a == b && c < d)
```

If you let `a=0`, `b=1` and `c > d` your statement is `true` and the original is `false`.

Reply:
They are the same condition, you can whittle the original down to this:

```
a < b || (a <= b && C)
```

and your proposal is 

```
a < b || (a == b && C)
```

Given the pw conditional for comparing bytes

<img width="635" alt="Screen Shot 2022-10-17 at 9 56 26 AM" src="https://user-images.githubusercontent.com/24281431/196236155-6e810705-32d9-4b0e-9a5d-61ab1d217492.png">

Though, I'm still not sure why we want to do this? They don't really read differently to me, each has the same number of atomic conditions.

Reply:
Your conditionals in the current version of the PR read totally fine to me; I was being overly pedantic haha. Thanks for the proof of their equality above, too (been a while since I've seen notation for a piecewise notation of a conditional) 🧑‍🔧 .